### PR TITLE
Allow to skip specific namespaces

### PR DESCRIPTION
**Pull request by jslee02**
This addresses the requested feature in #226.

### CODE REVIEW

**Comment by jslee02** [comment]
@psigen Please review this PR when you have a chance. This change really helps for building dartpy without any manual changes required from the auto-generated bindings.

File: include/chimera/configuration.h [hunk 1]
```diff
@@ -217,6 +222,7 @@ class CompiledConfiguration
     std::vector<std::pair<const clang::QualType, YAML::Node>> types_;
     std::map<const clang::Decl*, YAML::Node> declarations_;
     std::set<const clang::NamespaceDecl*> namespaces_;
```
Comment (psigen) [inline include/chimera/configuration.h]:
```suggestion
     std::set<const clang::NamespaceDecl*> namespacesIncluded_;
```

File: include/chimera/configuration.h [hunk 2]
```diff
@@ -217,6 +222,7 @@ class CompiledConfiguration
...
+    std::set<const clang::NamespaceDecl*> namespacesToSkip_;
```
Comment (psigen) [inline include/chimera/configuration.h]:
```suggestion
    std::set<const clang::NamespaceDecl*> namespacesSuppressed_;
```

File: include/chimera/configuration.h [hunk 3]
```diff
@@ -141,6 +141,11 @@ class CompiledConfiguration
      */
     const std::set<const clang::NamespaceDecl*>& GetNamespaces() const;
```
Comment (psigen) [inline include/chimera/configuration.h]:
```suggestion
    const std::set<const clang::NamespaceDecl*>& GetNamespacesIncluded() const;
```

File: include/chimera/configuration.h [hunk 4]
```diff
@@ -141,6 +141,11 @@ class CompiledConfiguration
...
 
+    /**
+     * Returns list of namespace declarations that should be skipped.
+     */
+    const std::set<const clang::NamespaceDecl*>& GetNamespacesToSkip() const;
```
Comment (psigen) [inline include/chimera/configuration.h]:
```suggestion
    const std::set<const clang::NamespaceDecl*>& GetNamespacesSuppressed() const;
```

File: src/configuration.cpp [hunk 5]
```diff
@@ -206,14 +206,21 @@ chimera::CompiledConfiguration::CompiledConfiguration(
...
             {
                 std::string ns_str = it.first.as<std::string>();
                 auto ns = chimera::util::resolveNamespace(ci, ns_str);
                 if (ns)
                 {
-                    declarations_[ns] = it.second;
-                    namespaces_.insert(ns);
+                    if (it.second.IsNull())
+                    {
+                        namespacesToSkip_.insert(ns);
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
                        namespacesSuppressed_.insert(ns);
```

File: src/configuration.cpp [hunk 6]
```diff
@@ -206,14 +206,21 @@ chimera::CompiledConfiguration::CompiledConfiguration(
...
+                    }
+                    else
+                    {
+                        declarations_[ns] = it.second;
+                        namespaces_.insert(ns);
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
                        namespacesIncluded_.insert(ns);
```

File: src/configuration.cpp [hunk 7]
```diff
@@ -461,6 +468,13 @@ chimera::CompiledConfiguration::~CompiledConfiguration()
 void
 chimera::CompiledConfiguration::AddTraversedNamespace(const clang::NamespaceDecl* decl)
 {
+    // Skip namespaces that are defined as null in the configuration.
+    for (const auto &it : GetNamespacesToSkip())
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
    for (const auto &it : GetNamespacesSuppressed())
```

File: src/configuration.cpp [hunk 8]
```diff
@@ -481,6 +495,12 @@ chimera::CompiledConfiguration::GetNamespaces() const
     return namespaces_;
 }
 
+const std::set<const clang::NamespaceDecl*>&
+chimera::CompiledConfiguration::GetNamespacesToSkip() const
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
chimera::CompiledConfiguration::GetNamespacesSuppressed() const
```

File: src/configuration.cpp [hunk 9]
```diff
@@ -481,6 +495,12 @@ chimera::CompiledConfiguration::GetNamespaces() const
...
+{
+    return namespacesToSkip_;
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
    return namespacesSuppressed_;
```

File: src/configuration.cpp [hunk 10]
```diff
@@ -481,6 +495,12 @@ chimera::CompiledConfiguration::GetNamespaces() const
     return namespaces_;
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
     return namespacesIncluded_;
```

File: src/configuration.cpp [hunk 11]
```diff
@@ -512,6 +532,13 @@ clang::ASTContext &chimera::CompiledConfiguration::GetContext() const
 
 bool chimera::CompiledConfiguration::IsEnclosed(const clang::Decl *decl) const
 {
+    // Skip namespaces that are defined as null in the configuration.
+    for (const auto &it : GetNamespacesToSkip())
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
    for (const auto &it : GetNamespacesSuppressed())
```

File: src/configuration.cpp [hunk 12]
```diff
@@ -512,6 +532,13 @@ clang::ASTContext &chimera::CompiledConfiguration::GetContext() const
...
+    {
+        if (decl->getDeclContext() && it->Encloses(decl->getDeclContext()))
+            return false;
+    }
+
     // Filter over the namespaces and only traverse ones that are enclosed
     // by one of the configuration namespaces.
     for (const auto &it : GetNamespaces())
```
Comment (psigen) [inline src/configuration.cpp]:
```suggestion
     for (const auto &it : GetNamespacesIncluded())
```

**Comment by jslee02** [comment]
> It looks like we might be able to put this logic into the top of the `IsSuppressed(decl)` function, and then use that in the position of `IsEnclosed()`.

I like the idea that making `IsSuppressed()` to check suppressed namespace. But I'm not sure if you mean removing `IsEnclosed()` in favor of `IsSuppressed()`. Could you let me know what was your intention?

**Comment by psigen** [comment]
Sorry: here is what I think makes sense:
* Put the checks against both `GetNamespaceSuppressed()` and `GetNamespaceIncluded()` into `IsEnclosed()`, as it was before.
* Make `if (!IsEnclosed(decl)) { return true; }` be the first check of `IsSuppressed()` so we don't repeat the namespace enclosure checking logic.
* Call `IsSuppressed()` instead of `IsEnclosed()` in [visitor.cpp](https://github.com/personalrobotics/chimera/blob/master/src/visitor.cpp#L98).
* Remove other usages of `IsSuppressed()` (because I _think_ they will be superseded by the one initial call at `VisitDecl`).  **This is the step that might introduce issues.**

Does that make sense?  The only thing I am unsure of is whether rearranging these checks will have any weird side-effects.  I don't see why they should but I'm not sure.

**Comment by jslee02** [comment]
Thanks for the guide! I tested this on `dartpy`, and it worked as expected without introducing any side effects. Let me merge this!

**Pull request merged**